packages/sqlite-pool/src/__tests__/limited.test.ts

```diff
@@ -0,0 +1,45 @@
+import createConnectionPool, {sql} from '../';
+
+jest.setTimeout(30000);
+
+const db = createConnectionPool(
+  ':memory:',
+  {},
+  {maxSize: 2, releaseTimeoutMilliseconds: 100},
+);
+
+afterAll(async () => {
+  await db.dispose();
+});
+
+test('two parallel queries', async () => {
```

Comment:
@ForbesLindesay this is the use case I’m talking about. Note that the two transactions are running in parallel.

Note that the db access is still synchronous, but if other asynchronous actions (eg an http request) happen within the transaction, things can progress in parallel.

Reply:
I don't want to merge something that only adds this psudeo-asynchronousness, and does not add true parallel queries:

- Doing other async work inside a transaction is normally a very bad idea.
  - In SQLite it means that no other database transactions or queries that write data can run until that async work completes. You probably don't want to block all database writes while waiting for some external HTTP service.
  - Even in other database engines like Postgres, open transactions are an expensive/limited resource. Even very large database servers might be limited to a few hundred open connections/transactions before they stall.
  - A transaction cannot be truly atomic if it needs to call external services other than the database. If a later step in the transaction fails, there is no way to "roll back" the external API call. This can often lead to inconsistent states. It's therefore normally better to make a decision to either do the db transaction, then call the external service, or call the external service then do the db transaction. This limits your possible failure modes - you only need to consider what happens if the server crashes between those two steps.
- I'm concerned people would assume this is more parallel than it is. For example, people would assume that `await db.query(...)` would not block the main JavaScript thread, but that assumption is broken here. That's a big part of why I preferred a clearly synchronous API for `sqlite-sync` over this pseudo-async API in the first place.

This test also does not demonstrate that you can even run a read only query in parallel while there is an active open transaction. I do think that may be possible with this code, but I don't think this test actually demonstrates that functionality. Each of these transactions is simply run sequentially, one after another.

If you do still want to do this (and again, I don't personally think it's a good idea), you only actually need two connections: primary (for transactions and queries that write data) and secondary (for queries that only read data). You don't need any of the connection pooling logic because you are not able to run more than one "secondary" query at a time. I'm not sure about including this in the core `@databases` project though, due to the reasons above.

We could consider something like `@databases/sqlite-sync-compat` to be a package that provides an API that looks like the API for `@databases/pg`/`@databases/mysql`/`@databases/sqlite` but is not actually async and still just uses a single connection. You could remove all the connection pooling stuff and simply take a lock on the entire connection for the duration of any transactions. I'm still not sure this is a good idea/something that should be part of the `@databases` core project.

An important non-goal of `@databases` is being able to reuse code between different database engines. I aim to build best-in-class libraries for connecting to and querying each of the supported databases. I want developers who have used `@databases` to query one database engine to feel instantly familiar when querying another database engine and not have a significant learning curve to understand different APIs. That's why each library has such a similar structure. I do not want to hide genuine differences though. Many ORM layers can end up being limited to only supporting the feature set that's common to all database engines, or they have very inefficient implementations of some functionality on some database engines while being very fast on others. `@databases` libraries are specific to each database engine, meaning you get the best that that specific engine has to offer. In the case of SQLite via `better-sqlite3` that means a synchronous API, because the underlying database is not asynchronous. In the case of Postgres and MySQL that means connection pooling because the underlying database can support many concurrent connections/queries. My understanding of the platformatic project is that it's a higher level tool, so it may make sense for you to hide these differences between engines at that level, but I'm not sure `@databases` should do that.